Allow all on-prem users to access DB connector.

tarzzz opened the pull request.

tarzzz commented:
@chriddyp Instead of skipping authentication for on-premise, this simply adds the username to `ALLOWED_USERS` when they first authenticate through Plotly On-Prem Server, and `/users/current` also returns a valid username.. 

Please review..

chriddyp commented:
Looks good! Can you add a couple tests with `IS_RUNNING_INSIDE_ON_PREM=true`
- Check that username is added to the settings
- Check that auth works after logging in
- Check that auth doesn't work if user hasn't logged in
- Check that auth doesn't work if user's login failed

chriddyp commented on backend/routes.js:
```diff
@@ -279,7 +279,8 @@ export default class Servers {
 
-                    if (contains(username, getSetting('ALLOWED_USERS')) || that.isElectron || !getSetting('AUTH_ENABLED')) {
+                    if (contains(username, getSetting('ALLOWED_USERS')) || that.isElectron ||
+                        !getSetting('AUTH_ENABLED') || getSetting('IS_RUNNING_INSIDE_ON_PREM')) {
```
Some comments here that describe each of these conditions would be helpful

chriddyp commented on backend/routes.js:
```diff
@@ -279,7 +279,8 @@ export default class Servers {
                         status = 201;
```
could we move this `existingUsers.push({...})` to the next section after the `if (contains(username, getSetting('ALLOWED_USERS') || that.isElectron || ...)`? We don't use `existingUsers` in any other case and without reading the next section, it almost looks like we're accepting _any_ login, not just those that are authorized.

chriddyp commented on backend/routes.js:
```diff
@@ -279,7 +279,8 @@ export default class Servers {
-                    if (contains(username, getSetting('ALLOWED_USERS')) || that.isElectron || !getSetting('AUTH_ENABLED')) {
+                    if (contains(username, getSetting('ALLOWED_USERS')) || that.isElectron ||
+                        !getSetting('AUTH_ENABLED') || getSetting('IS_RUNNING_INSIDE_ON_PREM')) {
                         res.setCookie('plotly-auth-token', access_token, getCookieOptions());
```
Referring to https://github.com/plotly/falcon-sql-client/pull/213/files#r147733769, can we move `existingUsers.push(...)` and `status=201` here?

tarzzz commented:
@chriddyp Addressed comments..

  - Added new tests. Please note that I have merged the last two of the above mentioned test-scenarios into one.
  - Updated Auth related tests mentioned in: https://github.com/plotly/falcon-sql-client/pull/212#discussion_r147736949

tarzzz commented:
Update: Rebased to fix circleCI test failures. The commits are intact though.

chriddyp commented on backend/routes.js:
```diff
@@ -269,17 +269,16 @@ export default class Servers {
                     const existingUsers = getSetting('USERS');
                     const existingUsernames = pluck('username', existingUsers);
                     let status;
```
I guess we might as well move these 3 lines into the next section as well since they aren't used outside of that section's scope at all

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -331,13 +331,21 @@ describe('Authentication - ', () => {
+                    assert.deepEqual(json, {
+                        "PLOTLY_URL": "https://plot.ly",
+                        "USERS": ["plotly-database-connector"]
+                    });
```
👍  thanks!

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -377,6 +385,53 @@ describe('Authentication - ', () => {
+                assert.equal(res.status, 401);
+                done();
+            }).catch(done);
+        })
```
missing a `.catch(done)` here (and you can remove the `.catch(done)` above since you `return` the promise

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -377,6 +385,53 @@ describe('Authentication - ', () => {
+            return GET('settings').then(res => res.json().then(json => {
+                assert.equal(res.status, 200);
+                assert.deepEqual(json, {
+                    USERS: [ 'plotly-database-connector' ],
```
🐮  lint - no spaces around `[`

chriddyp commented on test/backend/routes.spec.js:
```diff
@@ -377,6 +385,53 @@ describe('Authentication - ', () => {
+                assert.deepEqual(getSetting('ALLOWED_USERS'), [username]);
+                done();
+            })).catch(done);
+        });
```
missing `.catch(done)` here (and you can remove the `.catch(done)` in the nested promise since you `return` it)

chriddyp commented:
a couple of more comments then 💃

tarzzz commented:
Updates made, merging !!
